h5netcdf/core.py

```diff
@@ -260,12 +273,32 @@ def __init__(self, parent, name):
                         self._determine_current_dimension_size(k, current_size)
 
                     self._dim_order[k] = dim_id
+                else:
+                    if self._root._phony_dim_count is not None:
+                        # if unscaled variable, get phony dimensions
+                        var_phony_dims = [v.shape[i] for i, j in
+                                          enumerate(v.dims) if len(j) == 0]
+                        phony_dims |= Counter(var_phony_dims)
+
                 if not _netcdf_dimension_but_not_variable(v):
                     var_name = k
                     if k.startswith('_nc4_non_coord_'):
                         var_name = k[len('_nc4_non_coord_'):]
                     self._variables.add(var_name)
 
+        # iterate over found phony dimensions and create them
+        if self._root._phony_dim_count is not None:
+            self._phony_dims = {}
+            for size in phony_dims:
+                for cnt in range(phony_dims[size]):
+                    name = 'phony_dim_{}'.format(
+                        len(self._phony_dims) +
+                        self._root._phony_dim_count)
+                    self._phony_dims[(size, cnt)] = name
+                    self._create_dimension(name, size)
+            # finally increase phony dim count
+            self._root._phony_dim_count += len(self._phony_dims)
```

Comment:
I notice that this keeps track of unique phony dimension name at the `File` rather than the `Group` level.

This seems reasonable, but just to verify, is this consistent with what netCDF4-Python does?

Reply:
AFAIK netcdf-c does all the hard work and netCDF4-Python just picks that up.  netcdf-c creates phony dims when iterating over all groups in a sorted manner. So yes, the current implementation mimicks the netcdf behaviour. 

Here we need to keep track of the count of the already created phony dimensions. This has to be on `File` level to get access from every subgroup.

I'll think about a test to compare netcdf4 and h5netcdf access.

Reply:
I guess this would be a pretty strange dataset, but you could make a file where two different groups have variables with unknown dimensions that appear in the different orders.